internal/unionstore/pipelined_memdb.go

```diff
@@ -0,0 +1,286 @@
+			p.onFlushing.Store(true)
+		}
+	}
+	p.flushingMemBuffer = p.memBuffer
+	p.len += p.flushingMemBuffer.Len()
+	p.size += p.flushingMemBuffer.Size()
+	p.memBuffer = newMemDB()
+	go func() {
+		util.EvalFailpoint("beforePipelinedFlush")
+		metrics.TiKVPipelinedFlushLenHistogram.Observe(float64(p.flushingMemBuffer.Len()))
+		metrics.TiKVPipelinedFlushSizeHistogram.Observe(float64(p.flushingMemBuffer.Size()))
+		err := p.flushFunc(p.flushingMemBuffer)
+		p.onFlushing.Store(false)
+		// Send the error to errCh after onFlushing status is set to false.
+		// this guarantees the onFlushing.Store(true) in another goroutine's MayFlush happens after onFlushing.Store(false) in this function.
+		p.errCh <- err
+	}()
+	return true, nil
+}
+
+// FlushWait will wait for all flushing tasks are done and return the error if there is a failure.
+func (p *PipelinedMemDB) FlushWait() error {
+	if p.flushingMemBuffer != nil {
+		err := <-p.errCh
+		// cleanup the flushingMemBuffer so the next call of FlushWait will not wait for the error channel.
+		p.flushingMemBuffer = nil
+		return err
+	}
+	return nil
+}
+
+// Iter implements the Retriever interface.
+func (p *PipelinedMemDB) Iter([]byte, []byte) (Iterator, error) {
+	return nil, errors.New("pipelined memdb does not support Iter")
+}
+
+// IterReverse implements the Retriever interface.
+func (p *PipelinedMemDB) IterReverse([]byte, []byte) (Iterator, error) {
+	return nil, errors.New("pipelined memdb does not support IterReverse")
+}
+
+// SetEntrySizeLimit sets the size limit for each entry and total buffer.
+func (p *PipelinedMemDB) SetEntrySizeLimit(entryLimit, bufferLimit uint64) {
+	p.memBuffer.SetEntrySizeLimit(entryLimit, bufferLimit)
+}
+
+// InspectStage implements MemBuffer interface.
+func (p *PipelinedMemDB) InspectStage(int, func([]byte, kv.KeyFlags, []byte)) {}
+
+// SnapshotIter implements MemBuffer interface.
+func (p *PipelinedMemDB) SnapshotIter(k, upperBound []byte) Iterator {
+	// TODO: return an iterator which output error.
+	return nil
+}
+
+// SnapshotIterReverse implements MemBuffer interface.
+func (p *PipelinedMemDB) SnapshotIterReverse(k, lowerBound []byte) Iterator {
+	return nil
+}
+
+// SnapshotGetter implements MemBuffer interface.
+func (p *PipelinedMemDB) SnapshotGetter() Getter {
+	return nil
+}
+
+func (p *PipelinedMemDB) Len() int {
+	return p.memBuffer.Len() + p.len
+}
+
+func (p *PipelinedMemDB) Size() int {
+	return p.memBuffer.Size() + p.size
+}
+
+func (p *PipelinedMemDB) OnFlushing() bool {
+	return p.onFlushing.Load()
+}
+
+// SetMemoryFootprintChangeHook sets the hook for memory footprint change.
+func (p *PipelinedMemDB) SetMemoryFootprintChangeHook(hook func(uint64)) {}
+
+// Mem returns the memory usage of MemBuffer.
+func (p *PipelinedMemDB) Mem() uint64 {
+	return 0
+}
+
+// Reset the PipelinedMemDB to init state.
+func (p *PipelinedMemDB) Reset() {
+	p.memBuffer.Reset()
+	p.flushingMemBuffer = nil
+	p.len, p.size = 0, 0
+	p.onFlushing.Store(false)
```

Comment:
Why the CAS operation is used in the `Flush` function but the `Store` is used directly here? I dont't quite get the strategy of the concurrency design and handling here.

If the `write` operations of `PipelinedMemDB` are designed to be handled under the protection of `mutex`, is the `CAS` operation unnecessary?

Reply:
Like `MemDB`, `PipelinedMemDB` also does not allow parallel read/write operations, but the flush function runs in another goroutine, the atomic variable `onFlushing` avoids data races.

I remove the CAS operation by abstracting `needFlush` function and replace it with `atomic.Load`, the atomic variable is for quick checking if there is a ongoing flush task.